velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.cpp

```diff
@@ -26,73 +29,93 @@
-// TODO: Support multiple S3FileSystem instances using a cache
-static std::shared_ptr<S3FileSystem> s3fs = nullptr;
+std::mutex s3fsmutex;
+/// Multiple S3 filesystems are supported.
+/// Key is the endpoint value specified in the config using hive.s3.endpoint.
+/// If the endpoint is empty, it will default to AWS S3.
+/// We cannot use folly::ConcurrentHashMap to store the filesystem instances.
+/// This is because we need to delete the filesystems during
+/// finalizeS3FileSystem and folly::ConcurrentHashMap does not guarantee the
+/// entries are cleared on invoking clear().
+static std::unordered_map<std::string, std::shared_ptr<FileSystem>> filesystems;
 
-std::function<std::shared_ptr<
-    FileSystem>(std::shared_ptr<const Config>, std::string_view)>
-fileSystemGenerator() {
-  static auto filesystemGenerator = [](std::shared_ptr<const Config> properties,
-                                       std::string_view filePath) {
-    folly::call_once(S3FSInstantiationFlag, [&properties]() {
-      std::shared_ptr<S3FileSystem> fs;
-      if (properties != nullptr) {
-        initializeS3(properties.get());
-        fs = std::make_shared<S3FileSystem>(properties);
-      } else {
-        auto config = std::make_shared<core::MemConfig>();
-        initializeS3(config.get());
-        fs = std::make_shared<S3FileSystem>(config);
-      }
-      s3fs = fs;
-    });
-    return s3fs;
-  };
-  return filesystemGenerator;
-}
-
-std::function<std::unique_ptr<velox::dwio::common::FileSink>(
-    const std::string&,
-    const velox::dwio::common::FileSink::Options& options)>
-s3WriteFileSinkGenerator() {
-  static auto s3WriteFileSink =
-      [](const std::string& fileURI,
-         const velox::dwio::common::FileSink::Options& options)
-      -> std::unique_ptr<dwio::common::WriteFileSink> {
-    if (isS3File(fileURI)) {
-      auto fileSystem =
-          filesystems::getFileSystem(fileURI, options.connectorProperties);
-      return std::make_unique<dwio::common::WriteFileSink>(
-          fileSystem->openFileForWrite(fileURI, {{}, options.pool}),
-          fileURI,
-          options.metricLogger,
-          options.stats);
+std::shared_ptr<FileSystem> fileSystemGenerator(
+    std::shared_ptr<const Config> properties,
+    std::string_view filePath) {
+  static folly::
+      ConcurrentHashMap<std::string, std::shared_ptr<folly::once_flag>>
+          s3InitiationFlags;
+  std::string s3Identity("aws-s3");
+  if (properties) {
+    std::shared_ptr<HiveConfig> hiveConfig = std::make_shared<HiveConfig>(
+        std::make_shared<core::MemConfig>(properties->values()));
+    auto endpoint = hiveConfig->s3Endpoint();
+    if (!endpoint.empty()) {
+      // The identity is the endpoint.
+      s3Identity = endpoint;
+    }
+  }
+  /// If the init flag for a given s3 endpoint is not found,
+  /// create one for init use. It's a singleton.
+  if (s3InitiationFlags.find(s3Identity) == s3InitiationFlags.end()) {
```

Comment:
is there a way to encapsulate this code a bit more? For example, this seems similar to the function registry in `expression/VectorFunction.cpp` can we use a similar pattern?